libp2p/src/builder.rs

```diff
@@ -327,16 +326,14 @@ mod tests {
     #[test]
     #[cfg(all(feature = "tokio", feature = "tcp", feature = "tls", feature = "yamux"))]
     fn tcp_bandwidth_logging() -> Result<(), Box<dyn std::error::Error>> {
-        let (builder, _logging) = SwarmBuilder::with_new_identity()
+        let _ = SwarmBuilder::with_new_identity()
             .with_tokio()
             .with_tcp(
                 Default::default(),
                 libp2p_tls::Config::new,
                 libp2p_yamux::Config::default,
             )?
-            .with_bandwidth_logging();
-
-        builder
+            .with_bandwidth_logging(&mut libp2p_metrics::Registry::default())
             .with_behaviour(|_| libp2p_swarm::dummy::Behaviour)
             .unwrap()
             .build();
```

Comment:
@thomaseizinger you will like this. `with_bandwidth_logging` no longer breaks the builder pattern flow.

Reply:
Yes, very nice! Is this compatible with what polkadot does or do they use bandwidth logging differently?

Reply:
Should we call it `with_bandwidth_metrics` instead and ship it as a new feature rather than breaking the old one?

Reply:
> Is this compatible with what polkadot does or do they use bandwidth logging differently?

Once we expose `prometheus_client` metrics only, it is no longer compatible with Polkadot's metric system, given that they use `prometheus` (old client) instead of `prometheus_client` (official new client).

> Should we call it `with_bandwidth_metrics` instead and ship it as a new feature rather than breaking the old one?

:+1: I think that is the way to go. Deprecate `with_bandwidth_logging` and `Transport::with_bandwidth_logging`. Introduce `with_bandwidth_metrics` and a `libp2p_metrics::bandwidth::Transport`.

Polkadot can either switch over to the new system, or (for now) copy the old `libp2p::bandwidth_logging` code.

Reply:
I wonder if we should provide some kind of compatibility layer using a trait? Like add a trait called `Registry` and implement that type for the `prometheus_client::Registry`. I am not sure how API compatible they are though.

It would probably be a large trait in terms of surface ..

Reply:
The new `Registry` abstraction would need to have a associated type for a counter, ...

I don't think more abstraction is the way to go. I would actually like to push in the opposite direction and depend more on `prometheus-client` directly, instead of only through `libp2p-metrics`.

Downside:
- Direct dependency on `prometheus-client`. Though arguably it is a small dependency (`dtoa`, `itoa`, `parking_lot`, `syn`).
- Not compatible with other metrics crates (e.g. open telemetry or the old `prometheus` crate).

Upsite:
- We can fully embrace metrics. E.g. instrument deep internals, instead of only instrumenting at the surface.
- The `libp2p-metrics` abstraction can go away, making the metrics story much easier.

Whatever direction we choose, I suggest we don't do it within this pull request.

Reply:
> Whatever direction we choose, I suggest we don't do it within this pull request.

Yeah that is fair. I think if we only deprecate `with_bandwidth_logging` now, users still have to time to migrate if they are depending on the old `prometheus` crate.

> I would actually like to push in the opposite direction and depend more on `prometheus-client` directly, instead of only through `libp2p-metrics`.

I think that can be useful yes. Related to that: If we end up landing https://github.com/libp2p/rust-libp2p/pull/4282 then instrumenting deep internals might not be as necessary, at least not for diagnosing problems.